pkg/client/userd/daemon/service.go

```diff
@@ -131,106 +131,102 @@ func (s *Service) configReload(c context.Context) error {
 	return client.Watch(c, func(ctx context.Context) error {
 		s.sessionLock.RLock()
 		defer s.sessionLock.RUnlock()
-		return s.session.ApplyConfig(ctx)
+		if s.session == nil {
+			return client.RestoreDefaults(c, false)
+		}
+		return s.session.ApplyConfig(c)
 	})
 }
 
 // ManageSessions is the counterpart to the Connect method. It reads the connectCh, creates
 // a session and writes a reply to the connectErrCh. The session is then started if it was
 // successfully created.
 func ManageSessions(c context.Context, si userd.Service) error {
-	// The d.quit is called when we receive a Quit. Since it
-	// terminates this function, it terminates the whole process.
 	wg := sync.WaitGroup{}
+	defer wg.Wait()
+
 	var s *Service
 	si.As(&s)
-nextSession:
+
 	for {
 		// Wait for a connection request
-		var cr *rpc.ConnectRequest
 		select {
 		case <-c.Done():
-			break nextSession
-		case cr = <-s.connectRequest:
-		}
-
-		var session userd.Session
-		var rsp *rpc.ConnectInfo
-
-		s.sessionLock.Lock() // Locked during creation
-		if c.Err() == nil {  // If by the time we've got the session lock we're cancelled, then don't create the session and just leave by way of the select below
-			if s.session != nil {
-				// UpdateStatus sets rpc.ConnectInfo_ALREADY_CONNECTED if successful
-				rsp = s.session.UpdateStatus(s.sessionContext, cr)
-			} else {
-				sCtx, sCancel := context.WithCancel(c)
-				sCtx = userd.WithService(sCtx, si)
-				sCtx, session, rsp = userd.GetNewSessionFunc(c)(sCtx, s.scout, cr)
-				if sCtx.Err() == nil && rsp.Error == rpc.ConnectInfo_UNSPECIFIED {
-					s.sessionContext = session.WithK8sInterface(sCtx)
-					s.session = session
-					if err := s.session.ApplyConfig(c); err != nil {
-						dlog.Warnf(c, "failed to apply config from traffic-manager: %v", err)
-					}
-					s.sessionCancel = func() {
-						sCancel()
-						<-session.Done()
-					}
-				} else {
-					sCancel()
-					s.sessionCancel = nil
-				}
+			return nil
+		case cr := <-s.connectRequest:
+			rsp := startSession(c, si, cr, &wg)
+			select {
+			case <-c.Done():
```

Comment:
do we need this line since there is already a check on the context on line #152?

Reply:
Yes, it's needed. Starting the session might take a while and the context might have been cancelled during that time (calling `quit -s` would do that, for instance).